File: mvp/test-fixtures/python/01-fastapi-backend/src/controllers/ProductController.py
from typing import List, Optional
from ..services.ProductService import ProductService
from ..models.Product import Product


class ProductController:
    """FastAPI controller for product endpoints"""
    
    def __init__(self, product_service: ProductService):
        self.product_service = product_service
    
    def get_all(self) -> dict:
        """GET /products - Get all products"""
        products = self.product_service.get_all_products()
        return {
            'products': [p.to_dict() for p in products],
            'count': len(products)
        }
    
    def get_by_id(self, product_id: int) -> dict:
        """GET /products/{id} - Get product by ID"""
        product = self.product_service.get_product_by_id(product_id)
        if product:
            return {'product': product.to_dict()}
        return {'error': 'Product not found'}, 404
    
    def create(self, name: str, price: float, description: str = "", stock: int = 0) -> dict:
        """POST /products - Create new product"""
        product = self.product_service.create_product(name, price, description, stock)
        return {
            'product': product.to_dict(),
            'message': 'Product created successfully'
        }
    
    def update_stock(self, product_id: int, quantity: int) -> dict:
        """PATCH /products/{id}/stock - Update product stock"""
        product = self.product_service.update_product_stock(product_id, quantity)
        if product:
            return {
                'product': product.to_dict(),
                'message': 'Stock updated successfully'
            }
        return {'error': 'Product not found'}, 404
    
    def delete(self, product_id: int) -> dict:
        """DELETE /products/{id} - Delete product"""
        success = self.product_service.delete_product(product_id)
        if success:
            return {'message': 'Product deleted successfully'}
        return {'error': 'Product not found'}, 404
    
    def get_available(self) -> dict:
        """GET /products/available - Get available products"""
        products = self.product_service.get_available_products()
        return {
            'products': [p.to_dict() for p in products],
            'count': len(products)
        }

File: test-fixtures/php/01-laravel-backend/src/Controllers/ProductController.php
<?php

namespace App\Controllers;

use App\Services\ProductService;
use App\Models\Product;

/**
 * Laravel-style controller for product endpoints
 */
class ProductController
{
    private ProductService $productService;
    
    public function __construct(ProductService $productService)
    {
        $this->productService = $productService;
    }
    
    /**
     * GET /products - Get all products
     */
    public function index(): array
    {
        $products = $this->productService->getAllProducts();
        return [
            'products' => array_map(fn($p) => $p->toArray(), $products),
            'count' => count($products)
        ];
    }
    
    /**
     * GET /products/{id} - Get product by ID
     */
    public function show(int $id): array
    {
        $product = $this->productService->getProductById($id);
        if ($product) {
            return ['product' => $product->toArray()];
        }
        return ['error' => 'Product not found'];
    }
    
    /**
     * POST /products - Create new product
     */
    public function store(array $data): array
    {
        $product = $this->productService->createProduct(
            $data['name'],
            $data['price'],
            $data['description'] ?? "",
            $data['stock'] ?? 0
        );
        return [
            'product' => $product->toArray(),
            'message' => 'Product created successfully'
        ];
    }
    
    /**
     * PATCH /products/{id}/stock - Update product stock
     */
    public function updateStock(int $id, int $quantity): array
    {
        $product = $this->productService->updateProductStock($id, $quantity);
        if ($product) {
            return [
                'product' => $product->toArray(),
                'message' => 'Stock updated successfully'
            ];
        }
        return ['error' => 'Product not found'];
    }
    
    /**
     * DELETE /products/{id} - Delete product
     */
    public function destroy(int $id): array
    {
        $success = $this->productService->deleteProduct($id);
        if ($success) {
            return ['message' => 'Product deleted successfully'];
        }
        return ['error' => 'Product not found'];
    }
    
    /**
     * GET /products/available - Get available products
     */
    public function available(): array
    {
        $products = $this->productService->getAvailableProducts();
        return [
            'products' => array_map(fn($p) => $p->toArray(), $products),
            'count' => count($products)
        ];
    }
}
